File: 11.container-with-most-water.rs
/*
 * @lc app=leetcode id=11 lang=rust
 *
 * [11] Container With Most Water
 */

// @lc code=start
impl Solution {
    pub fn max_area(height: Vec<i32>) -> i32 {
        use std::cmp;

        let mut left = 0;
        let mut right = height.len() - 1;
        let mut area = 0;
        while left < right {
            let n = (right - left) as i32;
            let m:i32;
            if height[left] < height[right] {
                m = height[left];
                left += 1;
            } else {
                m = height[right];
                right -= 1;
            }
            let a = m * n;
            area = if a > area { a } else { area};
        }

        area
    }
}
// @lc code=end

File: 11.container-with-most-water.c
/*
 * @lc app=leetcode id=11 lang=c
 *
 * [11] Container With Most Water
 */

// @lc code=start
int maxArea(int* height, int heightSize){
    int water = 0, *i = height, *j = i + heightSize - 1;
    while (i < j) {
        int h = *i < *j ? *i : *j;
        int w = h * (j - i);
        if (w > water) water = w;
        while (*i <= h && i < j) i++;
        while (*j <= h && i < j) j--;
    }

    return water;
}
// @lc code=end

